### PR TITLE
Book Allocation Problem- JAVA

### PR DESCRIPTION
**Pull request by shahina-bano**
Added code for Book Allocation in JAVA for issue #665

### CODE REVIEW

**Comment by shahina-bano** [comment]
please review it @raksha009 @rudrakshi99

File: Code/Java/Book_Allocation.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
+
+/* problem statement: Provided "n" number of books and "m" number of students. Books are arranged in increasing number of its pages. Each student is assigned to read pages in consecutive order.
```
Comment (raksha009) [inline Code/Java/Book_Allocation.java:2]:
```suggestion
/* Book Allocation Problem
Problem Statement: 
Provided "n" number of books and "m" number of students. Books are arranged in increasing number of its pages. Each student is assigned to read pages in consecutive order.
```

File: Code/Java/Book_Allocation.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
+
+/* problem statement: Provided "n" number of books and "m" number of students. Books are arranged in increasing number of its pages. Each student is assigned to read pages in consecutive order.
+                      Assign books in such a manner that maximum number of pages each student get to read is minimum.
+*/ 
+
+
+
```
Comment (raksha009) [inline Code/Java/Book_Allocation.java:7]: Remove extra lines - line 6 and 7.

File: Code/Java/Book_Allocation.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                      Assign books in such a manner that maximum number of pages each student get to read is minimum.
+*/ 
+
+
+
+
+import java.util.*;
+import java.lang.*;
+import java.io.*;
+// / Name of the class hs to be "Main" only if the class is public. */
```
Comment (raksha009) [inline Code/Java/Book_Allocation.java:12]: Remove this line.

File: Code/Java/Book_Allocation.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+/* Test Case 1: 
+            input -
+                n = 6
+                m = 3
+                a[] = {13, 28, 34, 40, 57, 85}
+             output - 97
+
+ Test Case 2:
+            input -
```
Comment (raksha009) [inline Code/Java/Book_Allocation.java:101]:
```suggestion
            Input -

```
Keep I and O capital of Input and Output

**Review by raksha009 — CHANGES_REQUESTED** [review]
Please edit the readme of java and not of cpp :)

**Comment by shahina-bano** [comment]
changes are done...plz review them @raksha009 @OjusWiZard

File: Code/Java/Book_Allocation.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+Problem Statement: 
+Provided "n" number of books and "m" number of students. Books are arranged in increasing number of its pages. Each student is assigned to read pages in consecutive order.
+Assign books in such a manner that maximum number of pages each student get to read is minimum.
+*/ 
+
+import java.util.*;
+import java.lang.*;
+import java.io.*;
+
+class bookallocation
```
Comment (raksha009) [inline Code/Java/Book_Allocation.java:11]:
```suggestion
class Book_Allocation {

```

**Review by raksha009 — CHANGES_REQUESTED** [review]

**Review by raksha009 — APPROVED** [review]
LGTM!

**Comment by Kavs-123** [comment]
Is the issue resolved @raksha009 @aaadddiii!
If not i will like to contribute for the same ..

**Comment by shahina-bano** [comment]
I have created the PR 1 week ago and got 1 approval too for the Java language, so kindly review it @shreyanspoddar

**Comment by shahina-bano** [comment]
@raksha009  plz review

File: Algorithm/Searching_Sorting/readme.md [hunk 6]
```diff
@@ -68,4 +68,4 @@ Following are the types of searches which we will be discussing in this book.
 * Shell Sort ----> [C++](/Code/C++/shell_sort.cpp)
 * Selection Sort ----> [C++](/Code/C++/selection_sort.cpp)
 * Tim Sort ----> [C++](/Code/C++/tim_sort.cpp)
-
+* Book Allocation Problem ----> [Java](/Code/Java/Book_ALlocation.java)
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/readme.md:71]: It has to be Book Allocation only.

File: Code/Java/Book_Allocation.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+            int cnt = 1;
+
+            for(int j = 0; j < n; j++){
+                
+                // if number of pages is greater than key then it can't be allocated
+                if (a[j] > key) return false; 
+                
+                // counting the number of students required to allocate "sum" number of pages
+                if(sum + a[j] > key) {
+                    cnt++;       // incrementing the number students
```
Comment (raksha009) [inline Code/Java/Book_Allocation.java:24]: Add comments above the line.
Also, add appropriate comments only.

File: Code/Java/Book_Allocation.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+                     if(cnt > m) return false;   //if count of students is greater than given students "m", then its not possible to allocate these many books to given students "m"
+                }
+                else{
+                    sum += a[j];  // else update current number of pages
+                }
+            }
+            return true;
+    }
+        //function to find the page distribution
+    public static int binary_search( int a[], int n, int m){
```
Comment (raksha009) [inline Code/Java/Book_Allocation.java:36]: Indentation!

File: Code/Java/Book_Allocation.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+            int sum = 0;
+   
+            //counting total number of pages
+              for (int i = 0; i < n; i++) 
+                     sum += a[i]; 
+
+             int low = 0;
+             int high = sum;
+            // applying binary search for finding the minimum number of pages
+            while(low <= high){
```
Comment (raksha009) [inline Code/Java/Book_Allocation.java:51]: Indentation!

File: Code/Java/Book_Allocation.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+ Test Case 2:
+            Input -
+                 7
+                 5
+                 10 24 33 49 59 66 107
+            Output -
+                 Minimum number of pages = 107
+                
+            
+Time complexity : O(n*log(sum)) -  n = number of books, sum = sum of the pages of all the books
```
Comment (raksha009) [inline Code/Java/Book_Allocation.java:102]: where n = number of books

**Review by raksha009 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**